relay-server/src/utils/profile.rs

```diff
@@ -57,28 +76,140 @@ impl AndroidProfile {
     }
 }
 
-#[derive(Debug, Fail)]
-pub enum ProfileError {
-    #[fail(display = "invalid json in profile")]
-    InvalidJson(#[cause] serde_json::Error),
-    #[fail(display = "invalid base64 value")]
-    InvalidBase64Value,
-    #[fail(display = "invalid sampled profile")]
-    InvalidSampledProfile,
-    #[fail(display = "cannot serialize payload")]
-    CannotSerializePayload,
-}
-
 pub fn parse_android_profile(item: &mut Item) -> Result<(), ProfileError> {
     let mut profile: AndroidProfile =
         serde_json::from_slice(&item.payload()).map_err(ProfileError::InvalidJson)?;
 
     profile.parse()?;
 
+    if profile.profile.as_ref().unwrap().events.len() < 2 {
+        return Err(ProfileError::NotEnoughSamples);
+    }
+
+    match serde_json::to_vec(&profile) {
+        Ok(payload) => item.set_payload(ContentType::Json, &payload[..]),
+        Err(_) => return Err(ProfileError::CannotSerializePayload),
+    };
+
+    Ok(())
+}
+
+fn strip_pointer_authentication_code<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    let address_without_prefix = s.trim_start_matches("0x");
+    match u64::from_str_radix(address_without_prefix, 16) {
+        // https://github.com/microsoft/plcrashreporter/blob/748087386cfc517936315c107f722b146b0ad1ab/Source/PLCrashAsyncThread_arm.c#L84
+        Ok(address) => Ok(format!("{:#x}", address & 0x0000000FFFFFFFFF)),
+        Err(err) => Err(de::Error::custom(format!(
+            "Failed to strip pointer authentication code: {}",
+            err
+        ))),
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Frame {
+    function: String,
+
+    #[serde(deserialize_with = "strip_pointer_authentication_code")]
+    instruction_addr: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Sample {
+    frames: Vec<Frame>,
+    queue_address: Option<String>,
+
+    #[serde(deserialize_with = "deserialize_number_from_string")]
+    relative_timestamp_ns: u64,
```

Comment:
What do you mean by roundtrips? I don't serialize as string since I want to pass it as an integer downstream. Python supports decoding a `uint64` from a JSON object even if it's not in the standard.

I think I'd rather deserialize as an integer pass it downstream than use a float. This seems simpler as we're not going to have anymore type conversion.

Reply:
OK, I understand. The deserialization function I'm using now supports deserializing from string or number now.

Reply:
ok you're right. I think I'm fine with this then.